Implement transaction isolation level for SQLite

volka opened the pull request:
Implementation of isolation levels for start_transaction, see the pull request 

https://github.com/rbock/sqlpp11/pull/171

volka commented:
Ok the build seems to fail as it checks out the `develop` branch, which does not have the isolation_level merge yet. I'd not like to change the travis/appveyor scripts, could you perhaps do the merge?

rbock commented:
Did you see my comment on https://github.com/rbock/sqlpp11/pull/171 ?

I wonder whether passing the isolation level in the `start_transaction` call is a good idea? Reading the sqlite3 documentation, it seems to me that this will influence future transactions as well. Thus, it would be clearer to call `set_isolation_level()` instead.

What do you think?

volka commented:
Yes I know for Sqlite it is not ideal, as the level is indeed set per connection. I wanted to keep the interface similar for all databases as most do it per transaction (often with the option to set a connection/global default).

See my pull request for the PostgreSQL connector:
https://github.com/matthijs/sqlpp11-connector-postgresql/pull/28/files

PostgreSQL does it by transaction, allowing to set a default. MySQL allows setting the level by transaction, session/connection or globally.

To solve this for SQLite, I see three options for the current interface:

1. query the current level on transaction start and reset it in the destructor (more overhead for each transaction)
2. query the default level once at connection start, and set/reset in the transaction constructor/destructor only if it differs from the default (I would prefer this)
3. assume serializable (the "factory default") and only change when read_uncommitted is explicitly requested (least overhead)

or I would also be ok with implementing a set_isolation_level() in all connectors and just always set the default.

What would you prefer? Let me know and I will change the implementation accordingly.

rbock commented:
Hi Volker,

I like the "set_isolation_level() in all connectors and just always set the default" best. It seems to be the least surprising way across connectors.

Best,

Roland

volka commented:
OK, here we go. I removed the isolation_level parameter from start_transaction and added two functions "get/set_default_isolation_level()". I added the 'default' as for most databases, this will only affect new transactions but not the current one, and I hope the "default" will prompt people to look it up if in doubt ;)

I left the option for connectors to implement start_transaction(isolation_level) in addition to the global functions if supported (see the PostgreSQL change).

Is this better?

Regarding the connector api change in SQLPP11 itself, can I create the pull request against the develop branch now or will that create a mess with the already commited change in master? (change is here: https://github.com/volka/sqlpp11/commit/a55d986def7b3bc83aea42a5c46893ed1dcfba70)

rbock commented on include/sqlpp11/sqlite3/dynamic_libsqlite3.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015, Volker Aßmann
+ * Copyright (c) 2015, Volker A�mann
```
2017?

rbock commented:
Thanks for the change! I like the `default` in the name, too.

For sqlpp11, I think it would be easier to create a new pull request against master.

rbock commented:
Lost in isolation, but merged now :-)
